tests/run/i8726.scala

```diff
@@ -0,0 +1,5 @@
+case class A(a: Int)
+object C { def unapply(a: A): true | true = true }
+
+@main
+def Test =  (A(1): A | A) match { case C() => "OK" }
```

Comment:
In this case, do we optimize to not even call `C.unapply()`?

Reply:
Currently we don't. The potential optimization needs to assume properties of the method call, which can be performed in a later phase more generally.